### PR TITLE
chore: merge feature/components into feature/craft-application

### PR DESCRIPTION
**Pull request by sergiusens**
- [ ] Have you followed the [guidelines for contributing](https://github.com/snapcore/snapcraft/blob/master/CONTRIBUTING.md)?
- [ ] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [ ] Have you successfully run `tox run -m lint`?
- [ ] Have you successfully run `tox run -e test-py310`? (supported versions: `py39`, `py310`, `py311`, `py312`)

-----

### CODE REVIEW

File: snapcraft/providers.py [hunk 1]
```diff
@@ -197,15 +197,20 @@ def get_base_configuration(
 
     return bases.BuilddBase(
         alias=alias,
-        compatibility_tag=f"snapcraft-{bases.BuilddBase.compatibility_tag}.0",
+        compatibility_tag=f"snapcraft-{bases.BuilddBase.compatibility_tag}.1",
         environment=environment,
         hostname=instance_name,
         snaps=[
+            Snap(
```
Comment (mr-cal) [inline snapcraft/providers.py:204]: Should I make a new issue to only do this when `components` are defined?  
Currently, it is a subtask of https://github.com/canonical/snapcraft/issues/4644

It should be easy to implement (import craft_parts.features and check if partitions are enabled)

**Review by mr-cal — COMMENTED** [review]
Looks good.  The CI failures are due to https://github.com/canonical/snapcraft/pull/4651, but once https://github.com/canonical/craft-parts/pull/684 lands and you revert the changes to the requirement files here, the new top-level partition implementation will cause tests to start failing (like tests that assert for `prime/default/<file>`).

File: snapcraft/providers.py (hunk 1 above)
Comment (sergiusens) [reply snapcraft/providers.py:204]: I thought we would only merge here and hook up in a different task, so it seems appropriate

**Comment by sergiusens** [comment]
> Looks good. The CI failures are due to #465, but once [canonical/craft-parts#684](https://github.com/canonical/craft-parts/pull/684) lands and you revert the changes to the requirement files here, the new top-level partition implementation will cause tests to start failing (like tests that assert for `prime/default/<file>`).

Wrong issue linked?

**Comment by mr-cal** [comment]
> > Looks good. The CI failures are due to #465, but once [canonical/craft-parts#684](https://github.com/canonical/craft-parts/pull/684) lands and you revert the changes to the requirement files here, the new top-level partition implementation will cause tests to start failing (like tests that assert for `prime/default/<file>`).
> 
> Wrong issue linked?

Whoops, yes.  I corrected the original comment from `465` to `4651`

File: snapcraft/providers.py (hunk 1 above)
Comment (mr-cal) [reply snapcraft/providers.py:204]: Created: https://github.com/canonical/snapcraft/issues/4656

**Review by syu-w — APPROVED** [review]
Not sure if we need to wait the fixes

**Comment by tigarmo** [comment]
There are quite a few test failures

**Comment by tigarmo** [comment]
@sergiusens I see the component-related spread tests are still failing, do you want me to review the PR still?

**Comment by sergiusens** [comment]
@tigarmo organize is not organizing from the looks of it. Would be great if anyone can figure out why. I added a couple more test fixes and a spread cleanup as there were errors in how it was done (left overs)

**Comment by sergiusens** [comment]
we also do not need 3 tests for components, one could have been enough

**Review by tigarmo — APPROVED** [review]

File: tests/spread/core22/components/snapcraft.yaml [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+      # move to another partition
+      file-in-component: (component/share)/file-in-component
+      (default)/other-file-in-component: (component/share)/other-file-in-component
+
+    stage:
+      # exclude a file in the default partition
+      - -(default)/other-file-in-snap
+    prime:
+      # exclude a file in a partitionk
```
Comment (mr-cal) [inline tests/spread/core22/components/snapcraft.yaml:52]:
```suggestion
      # exclude a file in a partition
```

File: setup.py [hunk 3]
```diff
@@ -101,7 +101,7 @@ def recursive_data_files(directory, install_directory):
     "craft-cli",
     "craft-grammar",
     "craft-parts",
-    "craft-providers",
+    "craft-providers @ git+https://github.com/canonical/craft-providers@main#egg=craft-providers",
```
Comment (mr-cal) [inline setup.py:104]: craft-providers `1.23.1` is released, I don't think you need to bring this in

File: tests/spread/core22/components/task.yaml [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+prepare: |
+  snap refresh snapd --edge
+
+restore: |
+  snapcraft clean --destructive-mode
+  rm -f ./*.snap ./*.comp
+  snap refresh snapd --stable
+
+execute: |
+  # Craft Parts is not partitions aware during organize in managed mode
```
Comment (mr-cal) [inline tests/spread/core22/components/task.yaml:12]: https://github.com/canonical/craft-parts/pull/689 makes this test pass without needing `--destructive-mode`

**Review by mr-cal — APPROVED** [review]

File: setup.py (hunk 3 above)
Comment (sergiusens) [reply setup.py:104]: great! I thought I had gotten rid of this!

File: tests/spread/core22/components/task.yaml (hunk 4 above)
Comment (sergiusens) [reply tests/spread/core22/components/task.yaml:12]: might as well have @cmatsuoka release that tomorrow and I can just ensure we update the test here

**Pull request merged**